File: 0079-word-search/0079-word-search.go
func exist(board [][]byte, word string) bool {

	ROW := len(board)
	COL := len(board[0])
	visited := make(map[int]map[int]bool)

	var dfs func(row, col, index int) bool
	dfs = func(row, col, index int) bool {
		if index == len(word) {
			return true
		}

		value1, ok1 := visited[row]
		if ok1 {
			value2, ok2 := value1[col]
			if ok2 && value2 {
				return false
			}
		}

		if row < 0 || col < 0 || row >= ROW || col >= COL || board[row][col] != word[index] {
			return false
		}

        if len(visited[row]) == 0 {
			visited[row] = make(map[int]bool)
		}
		visited[row][col] = true
		result := dfs(row-1, col, index+1) || dfs(row+1, col, index+1) || dfs(row, col-1, index+1) || dfs(row, col+1, index+1)
		delete(visited[row], col)
		if len(visited[row]) == 0 {
			delete(visited, row)
		}

		return result
	}

	for r := range board {
		for c := range board[r] {
			if dfs(r, c, 0) {
				return true
			}
		}
	}
	return false
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        ROW, COL = len(board), len(board[0])
        visited = set()
        
        def dfs(r,c,i):
            if i == len(word):
                return True
            
            if (r < 0 or c < 0 or r >= ROW or c >= COL or board[r][c] != word[i] or (r,c) in visited):
                return False
            
            visited.add((r,c))
            res = ( dfs(r+1,c,i+1) or
                  dfs(r-1,c,i+1) or
                  dfs(r,c+1,i+1) or
                  dfs(r,c-1,i+1) )
            visited.remove((r,c))
            
            return res
        
        for r in range(ROW):
            for c in range(COL):
                if dfs(r,c,0):
                    return True
                
        return False